### PR TITLE
Parser: Handle EGCs correctly and add support for string interpolation

### PR DESCRIPTION
**Pull request by OceanOak**
No changelog

### CODE REVIEW

**Comment by OceanOak** [comment]
Todo: 
- [ ] fix semantic tokenization  
- [ ] add tests for a String with significant whitespace (including newlines, spaces, and tabs)

File: backend/src/BuiltinExecution/Libs/Parser.fs [hunk 1]
```diff
@@ -29,6 +29,12 @@ let fns : List<BuiltInFn> =
       fn =
         (function
         | _, _, [ DString sourceCode ] ->
+          // This function converts a byte index into a character index. It was added to handle EGCs correctly
```
Comment (StachuDotNet) [inline backend/src/BuiltinExecution/Libs/Parser.fs:32]:
```suggestion
          // This was added to handle EGCs correctly
```
(the functionality is implied by the fn name)

File: backend/testfiles/execution/stdlib/semanticTokenization.dark [hunk 2]
```diff
@@ -871,6 +871,19 @@ module TokenizeExpression =
...
+                                        (TokenType.String, (0L, 2L), (0L, 3L))
+                                        (TokenType.String, (0L, 3L), (0L, 4L))
+                                        (TokenType.String, (0L, 4L), (0L, 5L))
+                                        (TokenType.String, (0L, 5L), (0L, 6L))
+                                        (TokenType.String, (0L, 6L), (0L, 7L))
+                                        (TokenType.String, (0L, 7L), (0L, 8L))
+                                        (TokenType.Symbol, (0L, 8L), (0L, 9L))
+                                        (TokenType.VariableName, (0L, 9L), (0L, 13L))
+                                        (TokenType.Symbol, (0L, 13L), (0L, 14L))
+                                        (TokenType.Symbol, (0L, 14L), (0L, 15L)) ]
```
Comment (StachuDotNet) [inline backend/testfiles/execution/stdlib/semanticTokenization.dark:885]:
```suggestion
  ("$\"hello {name}\"" |> tokenize) = [ (TokenType.Symbol, (0L, 0L), (0L, 1L)) // $
                                        (TokenType.Symbol, (0L, 1L), (0L, 2L)) // "
                                        (TokenType.String, (0L, 2L), (0L, 3L)) // h
                                        (TokenType.String, (0L, 3L), (0L, 4L)) // e
                                        (TokenType.String, (0L, 4L), (0L, 5L)) // l
                                        (TokenType.String, (0L, 5L), (0L, 6L)) // l
                                        (TokenType.String, (0L, 6L), (0L, 7L)) // o
                                        (TokenType.String, (0L, 7L), (0L, 8L))
                                        (TokenType.Symbol, (0L, 8L), (0L, 9L))
                                        (TokenType.VariableName, (0L, 9L), (0L, 13L))
                                        (TokenType.Symbol, (0L, 13L), (0L, 14L))
                                        (TokenType.Symbol, (0L, 14L), (0L, 15L)) ]
```

This looks wrong - I made a few comments, but this breaks the tokenization _many_ individual tokens, rather than capturing `hello ` all at once.

Comment (StachuDotNet) [reply backend/testfiles/execution/stdlib/semanticTokenization.dark:885]: It makes me wonder if this string is being parsed with a _bunch_ of `StringSegment`s

Comment (OceanOak) [reply backend/testfiles/execution/stdlib/semanticTokenization.dark:885]: Not `StringSegment, it is capturing them as `string_text`
```
    string_interpolation_content: $ =>
      repeat1(choice($.string_to_eval, $.string_text)),

    string_text: $ =>
      prec.left(
        repeat1(
          choice(
            token.immediate(prec(1, /[^\\"{}\n]/)),
            $.char_or_string_escape_sequence,
          ),
        ),
      ),
```

couldn't really find a way to capture it as the full thing and at the same time keep the order of `string_text` and `string_to_eval` correct

File: packages/darklang/languageTools/writtenTypesToProgramTypes.dark [hunk 3]
```diff
@@ -344,6 +344,19 @@ module Darklang =
...
+          (owner: String)
+          (currentModule: List<String>)
+          (segment: WrittenTypes.StringSegment)
+          : ProgramTypes.StringSegment =
+          match segment with
+          | StringText(_, s) -> ProgramTypes.StringSegment.StringText s
+          | StringInterpolation(_, expr, _, _) ->
+            ProgramTypes.StringSegment.StringInterpolation(
+              Expr.toPT onMissing pm owner currentModule expr
+            )
```
Comment (StachuDotNet) [inline packages/darklang/languageTools/writtenTypesToProgramTypes.dark:359]: might be cleaner to inline this in the `Expr.toPT`
- LetPattern is simple, but used by both ELambda and ELet, so extracted is good
- MatchPattern and PipeExpr are only used in one place each (I think) but are more complicated so good to extract

File: tree-sitter-darklang/grammar.js [hunk 4]
```diff
@@ -442,11 +442,44 @@ module.exports = grammar({
...
+      ),
+
+    string_to_eval: $ =>
+      seq(
+        field("symbol_open_brace", alias("{", $.symbol)),
+        field("expr", $.expression),
+        field("symbol_close_brace", alias("}", $.symbol)),
+      ),
+
+    string_segment: $ => choice($.string_literal, $.string_interpolation),
```
Comment (StachuDotNet) [inline tree-sitter-darklang/grammar.js:481]: Hmm, maybe `string_expr` would be a better name, since `string_segment` is already a concept

Comment (OceanOak) [reply tree-sitter-darklang/grammar.js:481]: in PT we call it `StringSegment` should we rename that as well?

File: backend/testfiles/execution/stdlib/semanticTokenization.dark (hunk 2 above)
Comment (StachuDotNet) [reply backend/testfiles/execution/stdlib/semanticTokenization.dark:885]: Hmm, I think it's worth figuring out what's going on here. If you want to merge and circle back to this, that's okay, but at some point we'll really want this to be one token. Imagine we have a _really long_ string, that involves a tiny bit of interpolation - that suddenly becomes a _lot_ of tokens. Plus, I think "being a single token" will influence your interactions in the editor - let's say I'm at `h|ello` and hit `ctrl+right` intending to jump to `hello|` - as-is, this might just jump to `he|llo`

File: tree-sitter-darklang/grammar.js (hunk 4 above)
Comment (StachuDotNet) [reply tree-sitter-darklang/grammar.js:481]: Hm, actually this is right - the thing I was thrown off by was that `string_segment` was right next to `int_literal` and such - was expecting _that_ to be something like `string_expr`.

File: backend/testfiles/execution/stdlib/semanticTokenization.dark (hunk 2 above)
Comment (OceanOak) [reply backend/testfiles/execution/stdlib/semanticTokenization.dark:885]: I agree, and I am not happy with this solution either, I'd prefer to look more into it before merging this.

File: packages/darklang/languageTools/parser/expr.dark [hunk 5]
```diff
@@ -130,17 +130,83 @@ module Darklang =
...
+              let findContents =
+                findAndParseOptional stringSegment "content" (fun stringPart ->
+                  (stringPart.range, stringPart.text) |> Stdlib.Result.Result.Ok)
+
+              let (range, contents) =
+                match findContents with
+                | Some contents ->
+                  let (range, contents) = contents
+                  (range, contents)
+                | None -> (stringSegment.range, "")
```
[the remark this answers is not in the document]
Comment (StachuDotNet) [reply packages/darklang/languageTools/parser/expr.dark:150]: ok to do in a later PR though

**Comment by StachuDotNet** [comment]
Nice!
